### PR TITLE
Change path to CMakeLists.txt to be relative to root, not pwd

### LINKED ISSUE #4383: CMake: root_cmakelists_dir Relative to Source
Migrated from #4374: the `CMakePackage` class currently assumes a user-defined `root_cmakelists_dir` is relative to `build_directory`.

More intuitive would be relative to `self.stage.source_path`.

a change (only) affects the packages: `bcl2fastq2` and `cpprestsdk`

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #4383 if everyone agrees that this is the right thing to do. I tried to make it behave similarly to `AutotoolsPackage`.

@ax3l Can you test this?

### CODE REVIEW

**Comment by ax3l** [comment]
thx, I will try to test it in the next days!

File: lib/spack/spack/build_systems/cmake.py [hunk 1]
```diff
@@ -83,14 +83,21 @@ def build_type(self):
         return 'RelWithDebInfo'
 
     @property
-    def root_cmakelists_dir(self):
-        """Returns the location of the root CMakeLists.txt
+    def cmakelists_directory(self):
+        """The relative path to the directory containing CMakeLists.txt
```
Comment (hartzell) [inline lib/spack/spack/build_systems/cmake.py]: It's not obvious to me what this path is relative *to*?  The staging directory or the expanded tarball or ....  Are there a few choice words that can be added to make it clear to those who aren't already in the know?

Comment (adamjstewart) [inline lib/spack/spack/build_systems/cmake.py]: "relative to the root directory of the extracted tarball"? I wasn't sure exactly what to call it. But I agree, the whole point of this PR is to make things clearer and more intuitive.

Comment (alalazo) [inline lib/spack/spack/build_systems/cmake.py]: I am not sure I find this name clearer than the previous one... I named the function `root_cmakelists_dir` because you may have more than one `CMakeLists.txt` file in your project, residing in different directories. Usually you want to point to the one in the root which will contain the `project` command, hence the name used before.

File: lib/spack/spack/build_systems/cmake.py [hunk 2]
```diff
@@ -83,14 +83,21 @@ def build_type(self):
...
 
-        :return: directory containing the root CMakeLists.txt
+        Defaults to the root of the extracted tarball.
+
+        :return: directory containing CMakeLists.txt
         """
         return self.stage.source_path
 
     @property
+    def cmakelists_abs_path(self):
```
Comment (alalazo) [inline lib/spack/spack/build_systems/cmake.py]: As we use this property in only one place, can we 'inline' it there?

File: lib/spack/spack/build_systems/cmake.py (hunk 1 above)
Comment (adamjstewart) [inline lib/spack/spack/build_systems/cmake.py]: I see. I renamed it to be more in line with what `AutotoolsPackage` calls it, but I could undo this change. Don't some Autotools projects also have multiple configure scripts? That might just be for vendored dependencies, but I know OpenMPI has a few.

File: lib/spack/spack/build_systems/cmake.py [hunk 3]
```diff
@@ -83,14 +83,21 @@ def build_type(self):
...
+        """Absolute path to the directory containing CMakeLists.txt."""
+        return os.path.abspath(self.cmakelists_directory)
```
Comment (ax3l) [inline lib/spack/spack/build_systems/cmake.py]: Error: `global name 'os' is not defined`
```
/home/src/spack/lib/spack/spack/build_systems/cmake.py:98, in cmakelists_abs_path:
     95       @property
     96       def cmakelists_abs_path(self):
     97           """Absolute path to the directory containing CMakeLists.txt."""
  >> 98           return os.path.abspath(self.cmakelists_directory)
```

Comment (adamjstewart) [inline lib/spack/spack/build_systems/cmake.py]: I'll fix that.

File: lib/spack/spack/build_systems/cmake.py [hunk 4]
```diff
@@ -103,13 +107,8 @@ def std_cmake_args(self):
     @staticmethod
     def _std_args(pkg):
         """Computes the standard cmake arguments for a generic package"""
-        try:
-            build_type = pkg.build_type()
-        except AttributeError:
-            build_type = 'RelWithDebInfo'
```
Comment (alalazo) [inline lib/spack/spack/build_systems/cmake.py]: @ax3l @adamjstewart This method was meant to be used *for a generic package*. Removing the try-except block makes it usable only for `CMakePackage` instances. The problem is that for backward compatibility reasons (not every package migrated to `CMakePackage` and a few of them support different build-systems) we are setting the attribute:
```python
# Standard CMake arguments
m.std_cmake_args = spack.CMakePackage._std_args(pkg)
```
on every package in Spack. Removing the block is very likely the cause of the error that @ax3l noticed in his review.

Comment (adamjstewart) [inline lib/spack/spack/build_systems/cmake.py]: Ah, I was wondering why that was there. I'll revert that change.

File: lib/spack/spack/build_systems/cmake.py [hunk 5]
```diff
@@ -143,8 +147,9 @@ def cmake_args(self):
 
     def cmake(self, spec, prefix):
         """Runs ``cmake`` in the build directory"""
-        options = [self.root_cmakelists_dir] + self.std_cmake_args + \
-            self.cmake_args()
+        options = [os.path.abspath(self.cmakelists_abs_path)]
```
Comment (ax3l) [inline lib/spack/spack/build_systems/cmake.py]: this var is not default-defined yet for `CMakePackage` objects :)

**Comment by adamjstewart** [comment]
Ping @tgamblin @alalazo. I can't keep rebasing this forever 😝
